### ISSUE TITLE
How to combine translation with customization?

### ISSUE THREAD

**Issue by TitovLab** [Ylianst/MeshCentral#837]
Our `agentinvite.handlebars` and `default.handlebars` files are slightly changed. To do this, we use a wonderful way to make changes by creating files in the `meshcentral-web/views` folder. This is very convenient when updating and familiar, since a similar mechanism is used in many programs, thank you!

How to combine the new Meshcentral multilingual interface and customization using the meshcentral-web folder?
If we act as before and create only the modified `agentinvite.handlebars` and `default.handlebars` files, this lead to the malfunctioning of the multilingual interface.
Is there a mechanism that takes into account customization files from the `meshcentral-web` folder when translating?

I tried to supplement the translation with a built-in translation mechanism (/translator.htm page), but it seems that it does not take into account the presence of `meshcentral-web` folder?

**Comment by Ylianst** [comment]
Oh, this is a good request. I need at add support for translation of the "meshcentral-web/views" folder.

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.8-b with support for translated custom pages. Makes you changes to the English pages and place them in `meshcentral-web` as normal, then run `node node_modules/meshcentral --translate`. This will translate all you custom pages. You will then need to restart the server.

```
> node meshcentral --translate
Using default translate.json.
Translating HTML (cs): default.handlebars
Translating HTML (de): default.handlebars
Translating HTML (fr): default.handlebars
Translating HTML (ja): default.handlebars
Translating HTML (nl): default.handlebars
Translating HTML (pt): default.handlebars
Translating HTML (ru): default.handlebars
```

Note that if you have a "translate.json" file in the "meshcentral-data" folder, that translation dictionary will be used instead of the default one. Let me know if this works.

**Comment by TitovLab** [comment]
Unfortunately gives an error. Checked on the 0.4.8-i version.
MeshCentral is installed in `/opt/meshcentral/`, inside all folders, including `meshcentral-web` and `node_modules`.

Run `cd /opt/meshcentral/` and tried various options:
1. /usr/bin/node /opt/meshcentral/node_modules/meshcentral --translate
2. /usr/bin/node node_modules/meshcentral --translate
3. node node_modules/meshcentral --translate
4. exec /bin/su -c 'NODE_ENV=production /usr/bin/node /opt/meshcentral/node_modules/meshcentral --translate' -s /bin/bash meshcentral

Always the same error:
```
Using translate.json in meshentral-data.
internal/process/main_thread_only.js:25
    return binding.chdir(directory);
                   ^

Error: ENOENT: no such file or directory, chdir '/opt/meshcentral' -> './translate'
    at process.chdir (internal/process/main_thread_only.js:25:20)
    at process.chdir (/opt/meshcentral/node_modules/graceful-fs/polyfills.js:20:9)
    at Object.CreateMeshCentralServer.obj.Start (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:169:21)
    at /opt/meshcentral/node_modules/meshcentral/meshcentral.js:2342:110
    at InstallModules (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2244:121)
    at /opt/meshcentral/node_modules/meshcentral/meshcentral.js:2342:9
    at InstallModules (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2244:121)
    at mainStart (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2288:5)
    at Object.<anonymous> (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2358:5)
    at Module._compile (internal/modules/cjs/loader.js:816:30)
```

OS: CentOS 6 x64.

**Comment by TitovLab** [comment]
Updated to 0.4.8-m and created `/opt/meshcentral/translate/` folder.
And got new error:
```
[root@lin2 meshcentral]# node node_modules/meshcentral --translate
Using translate.json in meshentral-data.
Generating ../views/agentinvite-min.handlebars...
fs.js:119
    throw err;
    ^

Error: ENOENT: no such file or directory, open '../views/agentinvite.handlebars'
    at Object.openSync (fs.js:448:3)
    at Object.readFileSync (fs.js:348:35)
    at Object.startEx (/opt/meshcentral/node_modules/meshcentral/translate/translate.js:266:45)
    at Object.CreateMeshCentralServer.obj.Start (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:173:33)
    at /opt/meshcentral/node_modules/meshcentral/meshcentral.js:2350:110
    at InstallModules (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2250:121)
    at /opt/meshcentral/node_modules/meshcentral/meshcentral.js:2350:9
    at InstallModules (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2250:121)
    at mainStart (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2294:5)
    at Object.<anonymous> (/opt/meshcentral/node_modules/meshcentral/meshcentral.js:2366:5)
```
But file `/opt/meshcentral/meshcentral-web/views/agentinvite.handlebars` exist.
It seems that relative paths do not work?

**Comment by Ylianst** [comment]
Thanks for the detailed report. Working on this one now.

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.8-n that should fix this. The first time you run with --translate, it will install extra modules if needed. Let me know if this works.

**Comment by TitovLab** [comment]
Partially not working:
1. If there is a file `meshcentral-data/translate.json`, then:
1.1. **error**: in `meshcentral-web/views/` English `*-min.handlebars` are not created/updated
1.2. **error**: in `meshcentral-web/views/` nothing happens at all, no translations created/updated
1.3. OK: in `node_modules/meshcentral/views/` all English `*-min.handlebars` are recreated and all translations in `node_modules/meshcentral/views/translations/` are recreated
Log:
```
[root@lin2 meshcentral]# node node_modules/meshcentral --translate
Using translate.json in meshentral-data.
Generating ../views/agentinvite-min.handlebars...
Generating ../views/default-min.handlebars...
Generating ../views/default-mobile-min.handlebars...
Generating ../views/download-min.handlebars...
Generating ../views/error404-min.handlebars...
Generating ../views/error404-mobile-min.handlebars...
Generating ../views/login-min.handlebars...
Generating ../views/login-mobile-min.handlebars...
Generating ../views/terms-min.handlebars...
Generating ../views/terms-mobile-min.handlebars...
Generating ../views/xterm-min.handlebars...
Generating ../views/message-min.handlebars...
Generating ../views/messenger-min.handlebars...
Generating ../views/player-min.handlebars...
Translating HTML (cs): agentinvite.handlebars
Translating HTML (cs): default.handlebars
Translating HTML (cs): default-mobile.handlebars
Translating HTML (cs): download.handlebars
Translating HTML (cs): error404.handlebars
Translating HTML (cs): error404-mobile.handlebars
Translating HTML (cs): login.handlebars
Translating HTML (cs): login-mobile.handlebars
Translating HTML (cs): terms.handlebars
Translating HTML (cs): terms-mobile.handlebars
Translating HTML (cs): xterm.handlebars
Translating HTML (cs): message.handlebars
Translating HTML (cs): messenger.handlebars
Translating HTML (cs): player.handlebars
Translating HTML (de): agentinvite.handlebars
Translating HTML (de): default.handlebars
Translating HTML (de): default-mobile.handlebars
Translating HTML (de): download.handlebars
Translating HTML (de): error404.handlebars
Translating HTML (de): error404-mobile.handlebars
Translating HTML (de): login.handlebars
Translating HTML (de): login-mobile.handlebars
Translating HTML (de): terms.handlebars
Translating HTML (de): terms-mobile.handlebars
Translating HTML (de): xterm.handlebars
Translating HTML (de): message.handlebars
Translating HTML (de): messenger.handlebars
Translating HTML (de): player.handlebars
Translating HTML (es): agentinvite.handlebars
Translating HTML (es): default.handlebars
Translating HTML (es): default-mobile.handlebars
Translating HTML (es): download.handlebars
Translating HTML (es): error404.handlebars
Translating HTML (es): error404-mobile.handlebars
Translating HTML (es): login.handlebars
Translating HTML (es): login-mobile.handlebars
Translating HTML (es): terms.handlebars
Translating HTML (es): terms-mobile.handlebars
Translating HTML (es): xterm.handlebars
Translating HTML (es): message.handlebars
Translating HTML (es): messenger.handlebars
Translating HTML (es): player.handlebars
Translating HTML (fr): agentinvite.handlebars
Translating HTML (fr): default.handlebars
Translating HTML (fr): default-mobile.handlebars
Translating HTML (fr): download.handlebars
Translating HTML (fr): error404.handlebars
Translating HTML (fr): error404-mobile.handlebars
Translating HTML (fr): login.handlebars
Translating HTML (fr): login-mobile.handlebars
Translating HTML (fr): terms.handlebars
Translating HTML (fr): terms-mobile.handlebars
Translating HTML (fr): xterm.handlebars
Translating HTML (fr): message.handlebars
Translating HTML (fr): messenger.handlebars
Translating HTML (fr): player.handlebars
Translating HTML (ja): agentinvite.handlebars
Translating HTML (ja): default.handlebars
Translating HTML (ja): default-mobile.handlebars
Translating HTML (ja): download.handlebars
Translating HTML (ja): error404.handlebars
Translating HTML (ja): error404-mobile.handlebars
Translating HTML (ja): login.handlebars
Translating HTML (ja): login-mobile.handlebars
Translating HTML (ja): terms.handlebars
Translating HTML (ja): terms-mobile.handlebars
Translating HTML (ja): xterm.handlebars
Translating HTML (ja): message.handlebars
Translating HTML (ja): messenger.handlebars
Translating HTML (ja): player.handlebars
Translating HTML (nl): agentinvite.handlebars
Translating HTML (nl): default.handlebars
Translating HTML (nl): default-mobile.handlebars
Translating HTML (nl): download.handlebars
Translating HTML (nl): error404.handlebars
Translating HTML (nl): error404-mobile.handlebars
Translating HTML (nl): login.handlebars
Translating HTML (nl): login-mobile.handlebars
Translating HTML (nl): terms.handlebars
Translating HTML (nl): terms-mobile.handlebars
Translating HTML (nl): xterm.handlebars
Translating HTML (nl): message.handlebars
Translating HTML (nl): messenger.handlebars
Translating HTML (nl): player.handlebars
Translating HTML (pt): agentinvite.handlebars
Translating HTML (pt): default.handlebars
Translating HTML (pt): default-mobile.handlebars
Translating HTML (pt): download.handlebars
Translating HTML (pt): error404.handlebars
Translating HTML (pt): error404-mobile.handlebars
Translating HTML (pt): login.handlebars
Translating HTML (pt): login-mobile.handlebars
Translating HTML (pt): terms.handlebars
Translating HTML (pt): terms-mobile.handlebars
Translating HTML (pt): xterm.handlebars
Translating HTML (pt): message.handlebars
Translating HTML (pt): messenger.handlebars
Translating HTML (pt): player.handlebars
Translating HTML (ru): agentinvite.handlebars
Translating HTML (ru): default.handlebars
Translating HTML (ru): default-mobile.handlebars
Translating HTML (ru): download.handlebars
Translating HTML (ru): error404.handlebars
Translating HTML (ru): error404-mobile.handlebars
Translating HTML (ru): login.handlebars
Translating HTML (ru): login-mobile.handlebars
Translating HTML (ru): terms.handlebars
Translating HTML (ru): terms-mobile.handlebars
Translating HTML (ru): xterm.handlebars
Translating HTML (ru): message.handlebars
Translating HTML (ru): messenger.handlebars
Translating HTML (ru): player.handlebars
Processing HTML: agentinvite.handlebars
Processing HTML: default.handlebars
Processing HTML: default-mobile.handlebars
Processing HTML: download.handlebars
Processing HTML: error404.handlebars
Processing HTML: error404-mobile.handlebars
Processing HTML: login.handlebars
Processing HTML: login-mobile.handlebars
Processing HTML: terms.handlebars
Processing HTML: terms-mobile.handlebars
Processing HTML: xterm.handlebars
Processing HTML: message.handlebars
Processing HTML: messenger.handlebars
Processing HTML: player.handlebars
1531 strings in output file.
```
2. If there is no `meshcentral-data/translate.json` file, then:
2.1. **error**: in `meshcentral-web/views/` English `*-min.handlebars` are not created/updated
2.2. OK: translations are created in `meshcentral-web/views/translations/` for customized handlebars, including minify
2.3. OK: in `node_modules/meshcentral/views/` nothing happens
Log:
```
[root@lin2 meshcentral]# node node_modules/meshcentral --translate
Using default translate.json.
Translating HTML (cs): default.handlebars
Translating HTML (de): default.handlebars
Translating HTML (es): default.handlebars
Translating HTML (fr): default.handlebars
Translating HTML (ja): default.handlebars
Translating HTML (nl): default.handlebars
Translating HTML (pt): default.handlebars
Translating HTML (ru): default.handlebars
```
3. `Translate Server` button in `translator.htm` do not update translations in `meshcentral-web/views/translations/` folder. Perhaps this is due to paragraph 1?
4. Previously, `*-min.handlebars` files consisted of one line. Now this is not so, but maybe it is now normal?

**Comment by Ylianst** [comment]
Published MeshCentral v0.4.8-s with fixes to --translate. Let me know if it's ok.

**Comment by TitovLab** [comment]
Now the working version, thanks!
Just in case, I think it can confusing other users: the `Translate Server` button in `translator.htm` doesn’t do anything in `meshcentral-web/views/`, is that what it is intended to be?
For convenience, I designed the results in a table:
![image](https://user-images.githubusercontent.com/11898568/74048343-152e9680-49e3-11ea-8620-a514830105eb.png)
